### ISSUE TITLE
Pyqgis doc for QgsFeature.setAttribute() function should mention exception

### ISSUE THREAD

**Issue by louis2607** [qgis/QGIS#51276]
### What is the issue about the documentation?

I wanted to improve a python function I have to set from a dictionary of field name keys and attribute values all the attributes of a feature however my improvement to remove what I thought to be redundant code doesn't work even though everything in the documentation seems to indicate it should.

```python
def set_attributes(feat: QgsFeature, attrs: Dict[str, Any]) -> None:
    for field_name, value in attrs.items():
        if feat.fieldNameIndex(field_name) != -1:
            feat.setAttribute(field_name, value)
```
to 

```python
def set_attributes(feat: QgsFeature, attrs: Dict[str, Any]) -> None:
    for field_name, value in attrs.items():
        feat.setAttribute(field_name, value)
```
causes a KeyError exception when the field `field_name` can't be found

I expected this to work since the underlying C++ function is already supposed to be making this check ([QgsFeature::setAttribute](https://api.qgis.org/api/classQgsFeature.html#a0075d1f9c830d62c8fda6b843c307dd8))
```cpp
bool QgsFeature::setAttribute( const QString &name, const QVariant &value )
{
  int fieldIdx = fieldNameIndex( name );
  if ( fieldIdx == -1 )
    return false;
 
  d.detach();
  d->attributes[fieldIdx] = value;
  d->valid = true;
  return true;
}
```
and it is what the documentation says it is supposed to do ([QgsFeature.setAttribute() C++ API doc](https://api.qgis.org/api/classQgsFeature.html#ac3c903deca2bf6bd72a69de57d0f2ab4))

> `setAttribute(const QString & name, const QVariant & value)
>
> Insert a value into attribute, by field name.
>
> Returns `False` if field `name` could not be matched.
>
> Field map must be associated using `setFields()` before this method can be used.
>
> Calling this method will automatically set the feature as valid (see `isValid()`).
>
> ...

However the Python API doc for the same function ([QgsFeature.setAttribute() Python API doc](https://qgis.org/pyqgis/master/core/QgsFeature.html#qgis.core.QgsFeature.setAttribute)) has an additional raises in the function description which goes in direct opposition with the returns value
> `setAttribute(self, name: str, value: Any)
>
> Insert a value into attribute, by field name.
>
> Returns `False` if field `name` could not be matched.
>
> Field map must be associated using `setFields()` before this method can be used.
>
> Calling this method will automatically set the feature as valid (see `isValid()`).
>
> ...
>
> Raises : `KeyError ` – if the attribute `name` could not be converted to an index

So either the wrong C++ API call is made or the Python API tries to convert the name to an index before calling `QgsFeature::setAttribute` or the documentation is wrongly written (besides the styling breaking in the Python API doc).

In any case, it is really confusing and should be resolved IMO.

**Comment by nicogodet** [comment]
https://api.qgis.org/api/qgsfeature_8h_source.html#l00631

In Python, it raises a KeyError. Python API description is wrong "Returns False if field name could not be matched." should not be there.

**Comment by louis2607** [comment]
Yeah it is confusing because of the documentation, but the behavior is still weird since you are required to check that the field name is in the feature before calling `QgsFeature.setAttribute(name: str, value)` which then checks again. 
As it is, this behavior makes the function basically useless since you have to check if the name is in the fields of the feature with `QgsFeature.fieldIndexName()` which means you have the index of the field so you can use the faster int index variant `QgsFeature.setAttribute(field: int)`.

Just seems like a weird decision to raise an exception in this case on the Python side when it is a True / False on the C++ side.

**Comment by nicogodet** [comment]
You can wrap it in a Try/Except

**Comment by louis2607** [comment]
I am aware of how to go around exceptions, I am raising an issue on the documentation mainly and, at least in my opinion, a design choice I have trouble finding a reason for.

**Comment by m-kuhn** [comment]
The advantage of this is, that it helps debugging. An exception must consciously be handled or will make noise when it happens. A return value can silently be ignored.
